File: Leetcode/C++/EqualRowAndColumnPairs.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int equalPairs(vector<vector<int>> &grid)
    {
        map<vector<int>, int> counter;
        int row = grid.size(), col = grid[0].size(), ans = 0;
        for (vector<int> row : grid)
        {
            counter[row] += 1;
        }
        for (int i = 0; i < row; i++)
        {
            vector<int> column_list = {};
            for (int j = 0; j < col; j++)
            {
                column_list.push_back(grid[j][i]);
            }
            ans += counter[column_list];
        }
        return ans;
    }
};

int main()
{
    Solution sol;
    int n;
    cin >> n;
    vector<vector<int>> grid(n, vector<int>(n));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> grid[i][j];
        }
    }

    int result = sol.equalPairs(grid);
    cout << result << endl;
    return 0;
}

File: Leetcode/Java/EqualRowAndColumnPairs.java
package Java;
import java.util.*;


class Solution {
    public int equalPairs(int[][] grid) {
        Map<List<Integer>, Integer> rowCounter = new HashMap<>();
        int n = grid.length;
        int ans = 0;

        // Count frequency of each row
        for (int i = 0; i < n; i++) {
            List<Integer> row = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                row.add(grid[i][j]);
            }
            rowCounter.put(row, rowCounter.getOrDefault(row, 0) + 1);
        }

        // Compare with columns
        for (int col = 0; col < n; col++) {
            List<Integer> column = new ArrayList<>();
            for (int row = 0; row < n; row++) {
                column.add(grid[row][col]);
            }
            ans += rowCounter.getOrDefault(column, 0);
        }

        return ans;
    }
}

public class EqualRowAndColumnPairs {
    public static void main(String[] args) {
        int[][] grid = {
            {3, 2, 1},
            {1, 7, 6},
            {2, 7, 7}
        };

        Solution solution = new Solution();
        int result = solution.equalPairs(grid);
        System.out.println("Equal Row and Column Pairs: " + result);
    }
}
